packages/react-components/react-virtualizer/src/components/Virtualizer/useVirtualizer.ts

```diff
@@ -0,0 +1,438 @@
+import { useIntersectionObserver } from '../../hooks/useIntersectionObserver';
+import type { ReactNode } from 'react';
+import { useEffect, useRef, useState } from 'react';
+
+import type { VirtualizerProps, VirtualizerState } from './Virtualizer.types';
+import { resolveShorthand } from '@fluentui/react-utilities';
+import { flushSync } from 'react-dom';
+
+export function useVirtualizer_unstable(props: VirtualizerProps): VirtualizerState {
+  const {
+    itemSize,
+    numItems,
+    virtualizerLength,
+    children: renderChild,
+    getItemSize,
+    bufferItems = Math.round(virtualizerLength / 4.0),
+    bufferSize = Math.floor(bufferItems / 2.0) * itemSize,
+    intersectionObserverRoot,
+    axis = 'vertical',
+    reversed = false,
+    onUpdateIndex,
+    onCalculateIndex,
+  } = props;
+
+  // Tracks the initial item to start virtualizer at, -1 implies first render cycle
+  const [virtualizerStartIndex, setVirtualizerStartIndex] = useState<number>(-1);
+
+  // Store ref to before padding element
+  const beforeElementRef = useRef<Element | null>(null);
+
+  // Store ref to before padding element
+  const afterElementRef = useRef<Element | null>(null);
+
+  // We need to store an array to track dynamic sizes, we can use this to incrementally update changes
+  const childSizes = useRef<number[]>(new Array<number>(getItemSize ? numItems : 0));
+
+  /* We keep track of the progressive sizing/placement down the list,
+  this helps us skip re-calculations unless children/size changes */
+  const childProgressiveSizes = useRef<number[]>(new Array<number>(getItemSize ? numItems : 0));
+
+  // The internal tracking STATE for child array (updates when re-render required).
+  const [childState, setChildState] = useState<ReactNode[]>([]);
```

Comment:
since you always read and make updates the `childArray` you can avoid storing an entire copy in memory by convering this to a `forceUpdate` function

```tsx
const forceUpdate = React.useReducer(() => ({}), {})[1];
```